File: src/test/java/com/kshitij/pocs/scloud/product/controller/ProductControllerTest.java
package com.kshitij.pocs.scloud.product.controller;

import com.kshitij.pocs.scloud.product.model.Product;
import com.kshitij.pocs.scloud.product.service.ProductService;
import lombok.extern.slf4j.Slf4j;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.http.MediaType;
import org.springframework.test.context.junit4.SpringRunner;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;

import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.when;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@RunWith(SpringRunner.class)
@WebMvcTest(ProductController.class)
@Slf4j
public class ProductControllerTest {

    @Autowired
    private MockMvc mvc;

    @MockBean
    ProductService productService;

    @Test
    public void getProduct_shouldGetProduct() throws Exception {
        log.info("Checking the product");
        when(productService.getProductById(1L)).thenReturn(new Product(1L,"New Product",2L));
        mvc.perform(
                MockMvcRequestBuilders.get("/product/1").accept(MediaType.APPLICATION_JSON))
                .andExpect(status().isOk()).andExpect(jsonPath("id").isNumber());
    }

}

File: src/test/kotlin/kshitij/pocs/scloud/productservice/ProductControllerTest.kt
package kshitij.pocs.scloud.productservice

import org.hamcrest.CoreMatchers.isA
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.equalTo
import org.junit.Test
import org.junit.jupiter.api.Assertions.*
import org.junit.runner.RunWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.web.client.TestRestTemplate
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.test.context.junit4.SpringRunner

@RunWith(SpringRunner::class)
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
class ProductControllerTest{

    @Autowired
    private lateinit var testRestTemplate:TestRestTemplate

    @Test
    fun `Test if we get a response from Base controller`() {
        val response = testRestTemplate.getForEntity("/product/Message", String::class.java)
        assertNotNull(response)
        assertEquals(response?.statusCode,HttpStatus.OK)
        assertEquals(response?.body,"Back with the Message")
    }

    @Test
    fun `get product by id`(){
        val response = testRestTemplate.getForEntity("/product/1", String::class.java)
        assertThat(response?.statusCode,equalTo(HttpStatus.OK))
        //assertEquals(response?.body,ResponseEntity<Product>)
    }
}